### DISCUSSION TITLE
Agile Rates are Unknown

### DISCUSSION THREAD

**Discussion by SteveEggo** [BottlecapDave/HomeAssistant-OctopusEnergy#463] [Q&A] [answered]
I have installed HomeAssistant-OctopusEnergy. Sensors are showing for some numbers e.g. Current Rate. I have enabled octopus_energy_electricity_current_day_rates. When I click on the current day rates, entity history shows as unknown.
With lozzd/octopus-energy-rates-card installed, when I try to use the custom card setting the configuration as 
entity: sensor.octopus_energy_electricity_xxx_xxx_current_day_rates
type: custom:octopus-energy-rates-card
cols: 2
It shows "Cannot read properties of undefined" followed by the config details as above.
If I replace the entity current_day_rates with current_rate I see the same number in each 30min slot
I am on a fixed rate tariff with Octopus, if I run the curl command
curl -u "sk_live_xxx:" "https://api.octopus.energy/v1/products/AGILE-FLEX-22-11-25/electricity-tariffs/E-1R-AGILE-FLEX-22-11-25-A/standard-unit-rates/"
I get values for each 30 min period

I an struggling to find a reference to current day rates in the Octopus Rest API Documentation
Does anyone understand where I am going wrong ?

**Comment by BottlecapDave** [comment]
Hello, there are a few reasons why things aren't working as intended.

> When I click on the current day rates, entity history shows as unknown.

The new rate sensors will not show any data until the next time the rate information is refreshed. This is because the entity does not listen for the raised events until it is enabled (HA feature). These are usually populated within 30 minutes (as long as a connection of OE servers can be made)

> With lozzd/octopus-energy-rates-card installed, when I try to use the custom card setting the configuration as
entity: sensor.octopus_energy_electricity_xxx_xxx_current_day_rates
type: custom:octopus-energy-rates-card
cols: 2
It shows "Cannot read properties of undefined" followed by the config details as above.

To my knowledge the rates card has not been updated since the last minor release, and therefore still relies on the `current_rate` sensor to read the rates from.

> If I replace the entity current_day_rates with current_rate I see the same number in each 30min slot
I am on a fixed rate tariff with Octopus

Only the rate information associated with the tariff that you're on will be downloaded. Therefore, if you're on a fixed rate tariff, you will not get agile rates and therefore the agile rates card will be useless as they'll just show the same rates all day. These can become available if you use the [override entities](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#tariff-overrides), and you point it to an agile tariff, but the data will be a day late.

**Reply by SteveEggo** [reply]
BottlecapDave, thank you for your reply
Point 1. Octopus Agile rates update at 4PM, my HA system has been running since this morning. current day rates still shows as unknown, althout it looks as it is being updated every 30mins.

![Current Day Rates](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/27545906/1e22b851-ec89-4bb2-9c90-b9616bc8e702)

Point 2. If I understand you correctly octopus-energy-rates-card should be using current day rates.

Point 3. Is this also saying I cannot display Agile tariffs for today, only yesterday.

New Point 4. Can you explain how you map from current day rates to the Octopus Rest API call. If that is what you are calling.

**Reply by BottlecapDave** [reply, marked as answer]
> Octopus Agile rates update at 4PM, my HA system has been running since this morning. current day rates still shows as unknown, althout it looks as it is being updated every 30mins.

If you go to the [developer tools](https://my.home-assistant.io/redirect/developer_template/) and add the following template (adjusting to your sensor) your output is unknown?

```
{{ state_attr('event.octopus_energy_electricity_xxx_xxx_current_day_rates', 'rates')[-1] if state_attr('event.octopus_energy_electricity_xxx_xxx_current_day_rates', 'rates') is not none else none }}
```

If the sensor is stating that it's updating every 30 minutes, then this attribute should be populated. The attribute is however not available via the normal entity view. The history of the entity is when the event was last received. The attributes can be seen in the [states](https://my.home-assistant.io/redirect/developer_states/) view.

> If I understand you correctly octopus-energy-rates-card should be using current day rates.

The rates card has not been updated to use the new event sensors, so you need to use the [current rate sensor](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#current-rate) with it for the time being.

> Is this also saying I cannot display Agile tariffs for today, only yesterday.

This is based on you saying in your original post of "I am on a fixed rate tariff with Octopus", which implies you are not on an agile tariff. If this is the case, everything mentioned thus far will not provide you with agile rates. This is because, as mentioned, this integration provides data based on your current tariff, not just agile. Therefore, the event sensor will have your fixed rates (at 30 minute intervals) and the rates card will be next to pointless (which is what you said you saw in your original post). 

If you are wanting to see a different tariff, then you'll need to use the formentioned [override entities](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/_docs/entities/electricity.md#tariff-overrides), which allow you to see your spending on another tariff. This uses your previous consumption data and so all rate information is a day behind.

> Point 4. Can you explain how you map from current day rates to the Octopus Rest API call. If that is what you are calling.

I am calling the appropriate [rate endpoint](https://developer.octopus.energy/docs/api/#list-tariff-charges) dependng on your tariff (basically the endpoint you curled). I am calling this with a period from of yesterday from midnight (e.g. 2023-10-12T00:00:00+01:00) to midnight two days from now (e.g. 2023-10-14T00:00:00+01:00). The data returned by this endpoint is then expanded to be in 30 minute intervals. The current day rates entity/event is the result where the valid_from/to dates are between midnight today and midnight tomorrow.

**Reply by SteveEggo** [reply]
BottlecapDave, thank you for your patience in answering all of these points. I will make more progress when I have an Agile account.
I should also thank you for all the work you have put into this so far, it seems to be a very well constructed project. I will definitely sponsor you in the near future.

**Reply by BottlecapDave** [reply]
No problem. Hope the integration works better for you when you move to Agile
